### PR TITLE
optionally suppress verbose output of passed tests by VERBOSE=2

### PR DESCRIPTION
**Pull request by DDvO**
In particular when adding new test cases it typically happens that some of them fail while others pass.
When setting the environment variable `VERBOSE=1` (or `V=1`) for debugging, the output of all tests is shown.
It can be quite annoying and a waste of time to manually skip over (irrelevant) output of passed tests when one just wants to find out which test cases failed for what reason.
This PR adds a new verbosity level `2` that shows verbose output only for those tests that failed.
The implementation is done essentially in `util/perl/OpenSSL/Test.pm` by buffering `STDOUT` and `STDERR`.

This new option required a generalization of the verbosity levels passed by `test/runtests.pl`: so far it has supported only levels `0` and `1`. Now it passes any value given as the level to `Test.pm`. 
Note that according to https://metacpan.org/pod/release/LEONT/Test-Harness-3.42/lib/TAP/Harness.pm the underlying `TAP::Harness` already supports some more levels:
```
 1   verbose        Print individual test results to STDOUT. 
 0   normal
-1   quiet          Suppress some test output (mostly failures while tests are running).
-2   really quiet   Suppress everything but the tests summary.
-3   silent         Suppress everything.
```

### CODE REVIEW

**Comment by mspncp** [comment]
> I can be quite annoying and a wast of time to manually skip over (irrelevant) output of passed tests when one just wants to find out which test cases failed for what reason.

Hello David, 

just a simple side note (which does not imply any judgement pro or contra your proposal): My personal workaround is to simply search for the string "not ok" in the output.

File: test/run_tests.pl [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 # Recognise VERBOSE and V which is common on other projects.
 BEGIN {
-    $ENV{HARNESS_VERBOSE} = "yes" if $ENV{VERBOSE} || $ENV{V};
+    $ENV{HARNESS_VERBOSE} = $ENV{V} ? $ENV{V} : $ENV{VERBOSE}
```
Comment (levitte) [inline test/run_tests.pl]: An even easier way to express that is `$ENV{HARNESS_VERBOSE} = $ENV{V} // $ENV{VERBOSE};`

**Comment by DDvO** [comment]
@mspncp, sure, but this is not strictly fun nor particularly productive when you have some 500 test cases (which I do, adding high-level CLI and CA conformance tests for the upcoming submission of our CMP client).

File: test/run_tests.pl (hunk 1 above)
Comment (DDvO) [reply test/run_tests.pl]: Cool - I'll use that.

**Comment by levitte** [comment]
I *like* the idea...  however, I think it would be better of `run()` was entirely in control...  but given the current code, I understand if that isn't possible.  To remedy this, I made a small change (#6426), which allows `run()` to pass down options to the command CODEref.  With that, you should be able to replace this line:

``` perl
    my ($cmd, $display_cmd) = shift->(0);
```

with this:

``` perl
    my %runopts;
    # Make a default stdout.  If the user has passed a stdout option of their own,
    # that one will override this.
    %runopts{stdout} = __results_file("$test_name.log")
        if $ENV{HARNESS_ACTIVE} && $ENV{HARNESS_VERBOSE} == 2;
    my ($cmd, $display_cmd) = shift->(%runopts);
```

and you can then remove the logic you added in `__decorate_cmd()`

**Comment by mspncp** [comment]
I feel a little uncomfortable about the fact that now `V=2` is _less_ verbose than `V=1`. The opposite way around it would be more logical. @levitte would you consider it a breaking change if the meanings would be swapped, i.e. `V={1,2}` shows output of {failing,all} tests?

**Comment by levitte** [comment]
I have no problems with such a change per se, no.

Another thing is that documentation is needed in INSTALL, and an entry should be added in CHANGES.

**Comment by levitte** [comment]
I'm having som second thoughts... The purpose here is to show the output for anything that comes out as `not ok`, right? In that case, test commands like this will completely miss the mark:

``` perl
    ok(!run(...));
```

This problem just became a whole lot more difficult.

**Comment by mspncp** [comment]
My understanding of it is that the test _verdict_ (`ok` vs. `not ok`) should be shown for _all_ tests (since it is also shown for `V=0`), only the test _details_ should be affected by the verbosity flag.

**Comment by levitte** [comment]
Yes, but *when* do you want to show the details?

**Comment by richsalz** [comment]
We are in feature freeze.  I think that should include tests enhancements, unless they are adding *new* tests.

**Comment by levitte** [comment]
Not sure what you're trying to say there, @richsalz...  or rather, how that applies to this PR

**Comment by richsalz** [comment]
This change seems to be enabling new features to the test framework.  I think that should wait until after this release goes out.  This fixes no bugs, addresses no open issues, and *we're in feature freeze* and that should apply to the test output as well.  There hasn't been any request to have this kind of thing.

**Comment by levitte** [comment]
Regarding the issue at hand, I think `run()` is the wrong place to *display* the verbose output, at least if you only want to display the output in the `not ok` case.  If that is what you want, then we need to add a wrapper for `ok()`, `is()` and `isnt()`, something like this:

``` perl
sub __display_output {
    # do whatever is needed
}

sub ok {
    my $res = Test::More::ok(@_);
    __display_output() unless $res;
    return $res;
}

sub is {
    my $res = Test::More::is(@_);
    __display_output() unless $res;
    return $res;
}

sub isnt {
    my $res = Test::More::isnt(@_);
    __display_output() unless $res;
    return $res;
}
```

As to what @richsalz says, we're in a feature freeze.  I'm not sure I agree that this PR counts as a new feature (it could actually be useful to see things more clearly in our humongous build logs), but I'm not going to rush to merge this either, as it's not *that* important, and also since this is not quite ready yet.

**Comment by levitte** [comment]
Made this 'after 1.1.1'.  We shouldn't stop discussion the solution, though.

**Comment by DDvO** [comment]
When submitting this PR today I did not want to imply any target release - it is not very urgent to me. I thought it can be discussed here and I can improve it according to the feedback, and it then can be merged whenever felt most appropriate. The label added by @levitte looks perfect to me.

**Comment by DDvO** [comment]
I tentatively chose level `2` just for backward compatibility as I was not sure if there is the freedom to re-interpret the levels. I fully agree that a linear semantics: of the levels makes more sense. When changing it anyway, I'd suggest a few more levels.

0: total silence
1: test summary only
2: additionally test verdict, number, and name
3: additionally output of failed test cases
4: additionally output of passed test cases

We could also take over the mentioned convention used by `TAP::Harness`, which apart from the offset by 3 is similar (where I read the semantics of `-1` as equivalent to `3` in the list above:

-3   silent           - Suppress everything.
-2   really quiet  - Suppress everything but the tests summary.
-1   quiet            - Suppress some test output (mostly failures while tests are running).
 0   normal
 1   verbose        - Print individual test results to STDOUT.

Of course, the changes need to be documented, but so far I had not known exactly where to do that.

**Comment by DDvO** [comment]
@levitte, glad to hear that you like the idea, too. 

Unfortunately I am lost with what you meant above with using #6426. All I wanted to achieve with the logic in `__decorate_cmd()` is that when the 'half-verbose' level is chosen, STDOUT and STDERR get redirected to the tmp file as far as they are not already being redirected to some somewhere else.

It's certainly a good point to choose the name of the tmp file based on the test_name.

I also do not quite understand what a problem you see with `run` doing the deferred test printing.
So far the printing is done by the test case itself (e.g., by C's `printf`) when called by `run`, so in a sense also within `run`. Or do you aim at making the output behavior more general/configurable? This could also be done in a separate PR.

**Comment by levitte** [comment]
OK, a bit of an explanation with the example I mentioned:

``` perl
    ok(!run(...))
```

That check succeeds (and outputs `ok`) when the program being run fails, and fails (and outputs `not ok`) when the program being run succeeds. For those two cases and with this PR, the output will become this:

```
{command line} => 1
.... Program output
.... Program output
.... Program output
ok
```

or:

```
not ok
```

That's not really what you wanted, right?

**Comment by DDvO** [comment]
Thanks @levitte for the explanation; got it (I had been too much focused on using the `exit_checker`, not thinking of potential negation of the test exit code).

I've meanwhile corrected the implementation based on your suggestions (where the logic within the redefined `sub ok` etc. needed improvement).

**Comment by DDvO** [comment]
I tested ;-) the test framework verbosity with 
```
SRCTOP=. BLDTOP=. V=2 test/run_tests.pl test_out_option
```
which includes both `ok(run(...` and `ok(!run(...` 
also temporarily changing some of its test cases in `test_out_option.t` to make them fail.

**Comment by DDvO** [comment]
I've just done two further improvements:
* For each test case its command line is echoed before (and not after) the command output.
* after each test case a separator line is printed (given sufficient level of verbosity).

**Comment by DDvO** [comment]
I've just done two further improvements:
* For each test case its command line is echoed before (and not after) the command output.
* after each test case a separator line is printed (given sufficient level of verbosity).

**Comment by DDvO** [comment]
There is still one thing I would like to improve.
When a test failed, after the verdict two extra lines are printed, for example:
```
not ok 1 - invalid output path: .
#   Failed test 'invalid output path: .'
#   at /proj/openssl/master/util/perl/OpenSSL/Test.pm line 561.
```
I found that the redundant test name and irrelevant Perl caller info (file and line) are printed by `Test::Builder::ok_debug` but I did not find a way to suppress this.
For instance, `*{Test::Builder::ok_debug} = sub {};` does not help. Does anyone know a solution for this?

**Comment by DDvO** [comment]
Closing this outdated PR in favor of the just merged #12088, which supersedes it.

**Pull request closed without merging**